### PR TITLE
UI auth

### PR DESCRIPTION
**Pull request by serprex**
Implement the most primitive authentication scheme I could think of

Extending this can be done in future by introducing new supporting environment variables

If someone feels queasy about storing password in cookie, a 2nd env variable can be made, `PEERDB_PASSWORD_TOKEN`, which the cookie is set to instead of the password

On cloud we'll want to use `pbkdf2` or some other key derivation function so that we never store the password anywhere. So add a `PEERDB_PASSWORD_FUNCTION` variable & set `PEERDB_PASSWORD` to `{algo:'SHA512',iter:999999,salt:'random',auth:'result'}` & have login handler hash password accordingly. Managed service would handle setting these environment variables on customer's setup

### CODE REVIEW

File: ui/components/Password.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use client';
+import { Button } from '@/lib/Button';
+
+export default function Password() {
```
Comment (iskakaushik) [inline ui/components/Password.tsx:4]: wouldn't it be more idiomatic to do something like:

```
const [pass, setPass] = useState<string>("");
...
... password: pass
```

File: ui/app/api/login/route.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { cookies } from 'next/headers';
+
+export async function POST(request: Request) {
+  const { password } = await request.json();
+  if (process.env.PEERDB_PASSWORD !== password) {
+    return new Response(JSON.stringify({ error: 'wrong password' }));
+  }
+  cookies().set('password', password, {
```
[the remark this answers is not in the document]
Comment (serprex) [reply ui/app/api/login/route.ts:8]: Hashing a password doesn't really do anything. Usually you'd never store password, only store pbkdf2 or whatever scrypt you prefer. But the spec is to have password in environment variable so it's a little late for not storing password

For cloud I'm thinking there'll be PEERDB_PASSWORD being a json blob of hash/salt/iter/algo & another environment variable PEERRDB_PASSWORD_FUNCTION which transforms password. That way when managed service configures it won't be storing passwords anywhere

File: ui/components/Password.tsx (hunk 1 above)
Comment (serprex) [reply ui/components/Password.tsx:4]: Yes

File: ui/app/api/login/route.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (iskakaushik) [reply ui/app/api/login/route.ts:8]: Do browsers encrypt cookies by default? I thought it was a `HTTPS` thing. The password is the the env but env is not directly exposed to the user of PeerDB right?

[the remark this answers is not in the document]
Comment (serprex) [reply ui/app/api/login/route.ts:8]: Arguably you could say you don't want to store the password on the user's computer. In that case, you should be storing derived key on their system. Or, have an authentication token env variable that gets compared & is what is stored in cookie alongside password. Or, store auth token in database

I'll add a 2nd env variable PEERDB_PASSWORD_TOKEN which can optionally be specified to not store password on user's computer. That token can be rotated, after which user would have to login again to refresh their cookie

**Comment by iskakaushik** [comment]
redirect for when password is incorrect, saying something like “invalid password, redirect to the login page”
landing page
- if password not set show what we have today
- if password is set, take them to a new page which then redirects to login page, for now this page can just have a text field to enter the password and handle login response.
for the text field make the border not white.

File: ui/app/api/login/route.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pankaj-peerdb) [reply ui/app/api/login/route.ts:8]: we can store an encrypted token(jwt) which has password, expiry.  That way we dont need to manage rotation.

[the remark this answers is not in the document]
Comment (serprex) [reply ui/app/api/login/route.ts:8]: Discussed offline

File: ui/components/Password.tsx [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
-      <input id='password' type='password' />
+      {error && <div style={{
+        borderRadius:'8px',
+        fontWeight:'bold',
+        color:'#600',
+        backgroundColor:'#c66'
+        }}>{error}</div>}
+      Password: <TextField variant='simple' value={pass} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setPass(e.target.value)} />
       <Button
         onClick={() => {
```
Comment (iskakaushik) [inline ui/components/Password.tsx:19]: on success we should redirect to the landing page

File: ui/app/page.tsx [hunk 4]
```diff
@@ -9,7 +10,7 @@ export default function Home() {
     <Layout sidebar={<SidebarComponent />}>
       <LayoutMain alignSelf='center' justifySelf='center' width='xxLarge'>
         <Header variant='largeTitle'>PeerDB Home Page</Header>
-        <>PEERDB_PASSWORD <Password /></>
+        {cookies().get('auth') && <Logout />}
```
Comment (iskakaushik) [inline ui/app/page.tsx:13]: can we move logout button to the bottom of the sidebar?

**Comment by iskakaushik** [comment]
There are some UI refinements for a later PR, but this is looking good for now. LGTM!

**Review by iskakaushik — APPROVED** [review]
left one comment for your consideration, but overall good

File: ui/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import type { NextRequest } from 'next/server';
+
+import { NextResponse } from 'next/server';
+
+export default function middleware(req: NextRequest) {
```
[the remark this answers is not in the document]
Comment (serprex) [reply ui/middleware.ts:5]: I'll add `export const config` but for now the rest seems unnecessary

**Pull request merged**